src/include/migraphx/op/roialign.hpp

```diff
@@ -0,0 +1,323 @@
+                const type* bottom_data  = x.data();
+                const auto roi_batch_ind = batch_indices[n];
+                // Do not using rounding; this implementation detail is critical
+                float roi_start_w = static_cast<float>(roi[roi_s.index({n, 0})] * spatial_scale);
+                float roi_start_h = static_cast<float>(roi[roi_s.index({n, 1})] * spatial_scale);
+                float roi_end_w   = static_cast<float>(roi[roi_s.index({n, 2})] * spatial_scale);
+                float roi_end_h   = static_cast<float>(roi[roi_s.index({n, 3})] * spatial_scale);
+
+                // Force malformed ROIs to be 1x1
+                float roi_width =
+                    (roi_end_w - roi_start_w) > 1.0f ? (roi_end_w - roi_start_w) : 1.0f;
+                float roi_height =
+                    (roi_end_h - roi_start_h) > 1.0f ? (roi_end_h - roi_start_h) : 1.0f;
+                float bin_size_h = static_cast<float>(roi_height / pooled_height);
+                float bin_size_w = static_cast<float>(roi_width / pooled_width);
+
+                // We use roi_bin_grid to sample the grid and mimic integral
+                int64_t roi_bin_grid_h =
+                    (sampling_ratio > 0)
+                        ? sampling_ratio
+                        : static_cast<int64_t>(std::ceil(roi_height / pooled_height));
+                int64_t roi_bin_grid_w =
+                    (sampling_ratio > 0)
+                        ? sampling_ratio
+                        : static_cast<int64_t>(std::ceil(roi_width / pooled_width));
+
+                // we want to precalculate indices and weights shared by all channels,
+                // this is the key point of optimization
+                std::vector<pos_weight> pre_calc(roi_bin_grid_h * roi_bin_grid_w * pooled_width *
+                                                 pooled_height);
+                std::vector<int64_t> lens = {
+                    pooled_height, pooled_width, roi_bin_grid_h, roi_bin_grid_w};
+                std::vector<std::size_t> comp_lens(lens.begin(), lens.end());
+                shape comp_s{shape::float_type, comp_lens};
+                this->calc_pos_weight(height,
+                                      width,
+                                      comp_s,
+                                      roi_start_h,
+                                      roi_start_w,
+                                      bin_size_h,
+                                      bin_size_w,
+                                      roi_bin_grid_h,
+                                      roi_bin_grid_w,
+                                      pre_calc);
+
+                std::vector<int64_t> lens1 = {channels, pooled_height, pooled_width};
+                std::vector<std::size_t> comp_lens1(lens1.begin(), lens1.end());
+                shape comp_s1{migraphx::shape::float_type, comp_lens1};
+                std::vector<int64_t> vec_index(channels, 0);
```

Comment:
What is the reason for having this as a vector of 0s? And if it's always passing in the value 0, what is the purpose of the `index` parameter for `calc_pooling`?

Reply:
As you can see, the calc_pooling is called in the shape_for_each(), so index is 0 for the first time calc_pooling called, and for the next time, the initial value of index is not 0 anymore. It is increased in the calc_pooling func and index is a reference in the function calc_pooling.